[MRG + 1] Remove  pilutil from examples

jotasi opened the pull request:
#### Reference Issues/PRs
See discussion in #10502 


#### What does this implement/fix?
#10502 replaced `scipy.misc.imresize`, which was deprecated by SciPy, with the local copy `sklearn.externals._pilutil.imresize`. As pointed out by @jnothman and agreed on in the linked discussion, this usage of a private module in an example is non-ideal. So this PR replaces it again by a combination of `scipy.ndimage.filters.gaussian_filter` and `skimage.image.rescale`. The new dependency (scikit-image) is added to the README and to circleCI as pointed out by @lesteve in his instructions on how to proceed (Thank you for these, btw).


#### Any other comments?
 - The results differ slightly from the older version. I'll post a comparison later (or tomorrow).
 - To avoid the overwriting of `face` from a function to images of different resolution, I introduced new variables `orig_face`, `smoothened_face`, and `rescaled_face`. If preferred, I can also change it back, though.
 - Different from what was discussed in #10502, I did not use `skimage.filters.gaussian`. This wrapper around `scipy.ndimage.filters.gaussian_filter` changed names from `skimage.filter.gaussian_filter` to `skimage.filters.gaussian_filter` in version 0.11 and then to `skimage.filters.gaussian` in 0.12, while `skimage.filters.gaussian_filter` was deprecated and is removed in the current dev version 0.14. As the resulting convolution of `try: except ImportError:` blocks did not seem right for a minor part of a simple example, I went with `scipy.ndimage.filters.gaussian_filter` instead.
 - So far I have checked compatibility with the minimum scikit-image version (0.9.3) in the README only in a local docker container. I'll add a link to the circleCI-artifacts as soon as the builds are done.
 - BTW, while adding scikit-image I saw that the versions in circleCI (SciPy=0.14, Matplotlib=1.3) do not match the minima in the README (SciPy>=0.13.3, Matplotlib>=1.1.1). Is this difference intentional or should this be updated in the README at some point?
 - Finally, both examples raise a warning on my machine:
`matplotlib/contour.py:967: UserWarning: The following kwargs were not used by contour: 'contours'`
`plot_face_segmentation.py` also raises:
`sklearn/utils/graph.py:115: FutureWarning: Conversion of the second argument of issubdtype from int to np.signedinteger is deprecated. In future, it will be treated as np.int64 == np.dtype(int).type.`
Should I remove the offending parts as well?

Commit:
Changed from internal _pilutil to skimage/scipy

Commit:
Updated requirements for the examples

Commit:
Added scikit-image needed in examples to circleCI

Commit:
gaussian_filter doesn't change dtype

jnothman commented:
You really need to stop referring to skimage as scikit-learn here. It's just too funny. Above you propose we need a new dependency on our own library...

jotasi commented:
Oh, you're obviously right, shame on me!

jotasi commented:
Here are the links to the artifacts for python 2 ([`plot_face_segmentation.py`](https://83-77606471-gh.circle-artifacts.com/0/doc/auto_examples/cluster/plot_face_segmentation.html) and [`plot_face_ward_segmentation.py`](https://83-77606471-gh.circle-artifacts.com/0/doc/auto_examples/cluster/plot_face_ward_segmentation.html)) and python 3 ([`plot_face_segmentation.py`](https://84-77606471-gh.circle-artifacts.com/0/doc/auto_examples/cluster/plot_face_segmentation.html) and [`plot_face_ward_segmentation.py`](https://84-77606471-gh.circle-artifacts.com/0/doc/auto_examples/cluster/plot_face_ward_segmentation.html)). The links to the examples in the current documentation can be found [here](http://scikit-learn.org/stable/auto_examples/cluster/plot_face_segmentation.html#sphx-glr-auto-examples-cluster-plot-face-segmentation-py) and [here](http://scikit-learn.org/stable/auto_examples/cluster/plot_face_ward_segmentation.html#sphx-glr-auto-examples-cluster-plot-face-ward-segmentation-py). I also put example plots of the differences produced on my local machine below. I hope there are not too many images now. Differences arise for example depending on when the conversion to `float`s is done and there is also a slight dependence on the boundary conditions. Should I try to tweak the resizing/smoothing to get the results to match more closely or are the results good enough the way they are now?

The resulting differences between before and after this PR on my local machine for `plot_face_ward_segmentation.py` are similar to those between artifacts and docs:
Now:
![for_gh_ward_now](https://user-images.githubusercontent.com/10883104/35405631-0f86dff6-0207-11e8-8b70-f17fca0b7a78.png)
Before:
![for_gh_ward_after_fix](https://user-images.githubusercontent.com/10883104/35405641-19161802-0207-11e8-84ac-9ad0672a0390.png)

For `plot_face_segmentation.py`, the situation is somewhat different. The result of the artifacts is quite similar on my machine but looks very different (at least in case of the 'kmeans' spectal clustering) to the version in the docs which confused me at first but seems to be due to the fix in #9062. Nonetheless, there are still some less pronounced differences:
Now:
![for_gh_kmeans_now](https://user-images.githubusercontent.com/10883104/35405695-4d5ce0be-0207-11e8-8263-76e317b37fc6.png)
![for_gh_discretize_now](https://user-images.githubusercontent.com/10883104/35405699-4fd167a2-0207-11e8-862f-dffec42c7ae0.png)
After #9062:
![for_gh_kmeans_after_fix](https://user-images.githubusercontent.com/10883104/35405707-53fb83da-0207-11e8-958c-d6c601476343.png)
![for_gh_discretize_after_fix](https://user-images.githubusercontent.com/10883104/35405710-557fce14-0207-11e8-82da-63de6546c900.png)
Before #9062:
![for_gh_kmeans_before_fix](https://user-images.githubusercontent.com/10883104/35405718-5b3e30b6-0207-11e8-8d4d-03cc904e110e.png)
![for_gh_discretize_before_fix](https://user-images.githubusercontent.com/10883104/35405724-5f18afea-0207-11e8-9b93-13ab75d81690.png)

lesteve commented:
> BTW, while adding scikit-image I saw that the versions in circleCI (SciPy=0.14, Matplotlib=1.3) do not match the minima in the README (SciPy>=0.13.3, Matplotlib>=1.1.1). Is this difference intentional or should this be updated in the README at some point?

Good catch I updated the README.md for matplotlib and pushed in master (sorry you'll need to fix the conflicts). At the moment we decided to use the versions from Ubuntu 14.04 as our minimal dependencies. We could specify the numpy and scipy version in .circleci/config.yml, if you could do that in a separate PR, that would be great!

For the image comparison you need to compare to the dev doc and not stable. To make it easier to compare, this is better if you post one image for this PR and the same image from the dev doc and then go on to the next image. If you could edit your previous post, that'd be great.

About the image difference, I don't know. Maybe @glemaitre and @jmargeta have some idea since they work on #9062.

Other comments:
* it's fine using `scipy.ndimage.filters.gaussian_filter`
* not sure about the warnings, it looks like they should be addressed. If you could do that in a separate PR or open an issue, this would be great.

glemaitre commented:
> About the image difference, I don't know. Maybe @glemaitre and @jmargeta have some idea since they work on #9062.

yep this is due to the fix. I work better now :) The last example (with kmeans) shows the fix. We get cluster which are more consistent.

jmargeta commented:
I second what @glemaitre says.

@jotasi, it is always good to see it being reproduced, thank you!
The differences not only in graph construction, but also random seed initialization and image type can indeed cause noticeable differences in the boundaries.

It leads me to question whether this image is actually a good example for the technique for a image segmentation / superpixels task. In contrast to papers using similar techniques I do not find the current results particularly convincing (e.g. boundaries are jagged and do not follow the real objects' boundaries). Wouldn't an image with crisper boundaries show the potential more clearly?

lesteve committed:
Merge branch 'master' into remove__pilutil_from_examples

lesteve commented:
I fixed the conflicts. I think this PR should focus on removing `sklearn.externals._pilutil` usages from the examples. Whether there is better image we can use for this example can be done in a separate issue/PR (and if we go ahead and merge this PR, we can take use one of the scikit-image image)

jotasi commented:
Thanks for taking care of the conflicts @lesteve, and thanks everyone for the feedback. I've updated my comment above now comparing the 3 images 1-to-1 with the dev-docs. I'll add separate issues/PRs for the following things:

- Adding numpy/scipy versions to circleci
- Fixing the warnings raised in the example
- Checking for a better image

glemaitre commented on examples/cluster/plot_face_segmentation.py:
```diff
@@ -26,26 +26,30 @@
 
 import numpy as np
 import scipy as sp
+from scipy.ndimage.filters import gaussian_filter
```
I would use scikit image:

http://scikit-image.org/docs/dev/api/skimage.filters.html#skimage.filters.gaussian

jotasi replied:
I found that as well. The problem with it is that it changed both name and module at different times since version 0.9.3 (which is the default in Ubuntu trusty) and thus one would need multiple `try:... except ImportError:...` which did not seem right for a simple example to me. (See the 3rd additional comment in opening post of the PR).
Do you agree that I then should keep SciPys version or should I change it anyway?

glemaitre commented on examples/cluster/plot_face_segmentation.py:
```diff
@@ -26,26 +26,30 @@
-face = imresize(face, 0.10) / 255.
+# Applying a Gaussian filter for smoothing prior to down-scaling
+# reduces aliasing artifacts.
+smoothened_face = gaussian_filter(orig_face/255., sigma=(1/0.10-1)/2.)
```
Actually you don't need that:

http://scikit-image.org/docs/dev/api/skimage.transform.html#skimage.transform.rescale

You can turn `anti_aliasing=True` to get the same effect.

jotasi replied:
Actually, this is only in the current dev-version. In the [docs of the current stable version](http://scikit-image.org/docs/0.13.x/api/skimage.transform.html#skimage.transform.rescale) (0.13.1) it is not listed, yet. One might want to keep this in mind and change in the (rather far) future, though.
([See also discussion here](https://github.com/scikit-learn/scikit-learn/pull/10502#issuecomment-359487323))

glemaitre replied:
Agreed. In this case rewrite it as:

```python
smoothened_face = gaussian_filter(orig_face, sigma=4.5)
```

glemaitre commented on examples/cluster/plot_face_ward_segmentation.py:
```diff
@@ -18,30 +18,35 @@
-face = imresize(face, 0.10) / 255.
+# Applying a Gaussian filter for smoothing prior to down-scaling
+# reduces aliasing artifacts.
+smoothened_face = gaussian_filter(orig_face/255., sigma=(1/0.10-1)/2.)
```
see my previous comment

jotasi replied:
Same here

jotasi commented:
Thanks for reviewing @glemaitre. I've explained my reasoning to your comments inline. Please let me know if you want to change this anyway (or something else).

glemaitre commented:
I see the point regarding the filtering. Let's keep scipy since that scikit-image is not released.

glemaitre commented on examples/cluster/plot_face_segmentation.py:
```diff
@@ -26,26 +26,30 @@
+    orig_face = face(gray=True)
 except ImportError:
-    face = sp.face(gray=True)
+    orig_face = sp.face(gray=True)
```
to avoid using the `/ 255`, let's use make the following

```python
from skimage import img_as_float64

orig_face = img_as_float64(sp.face(gray=True))
```

glemaitre commented on examples/cluster/plot_face_segmentation.py:
```diff
@@ -26,26 +26,30 @@
+# Applying a Gaussian filter for smoothing prior to down-scaling
+# reduces aliasing artifacts.
+smoothened_face = gaussian_filter(orig_face/255., sigma=(1/0.10-1)/2.)
+rescaled_face = rescale(smoothened_face, 0.10, mode="reflect")
```
0.10 -> 0.1

glemaitre commented on examples/cluster/plot_face_ward_segmentation.py:
```diff
@@ -18,30 +18,35 @@
+    orig_face = face(gray=True)
 except ImportError:
-    face = sp.face(gray=True)
+    orig_face = sp.face(gray=True)
```
See previous comments

Commit:
Implement glemaitres suggestions

jotasi commented:
I've implemented your suggestions. My original reasoning for `sigma=(1/0.10-1)/2.` was that this is the way `sigma` will be computed by scikit-image's `rescale` with `antialiasing=True` in the future but I agree that this is not really helpful without an explanation and as this not the point of the example, I've dropped it. Concerning using `skimage.img_as_float64`, also good point! Btw. I chose `skimage.img_as_float` instead, as the former is again only available in the dev-version and `skimage.img_as_float` produces `float64` anyway (see [the docs](http://scikit-image.org/docs/0.13.x/api/skimage.html#utility-functions)). I hope that this covers everything. Otherwise, please let me know!

glemaitre commented:
> Concerning using skimage.img_as_float64, also good point! Btw. I chose skimage.img_as_float instead, as the former is again only available in the dev-version and skimage.img_as_float produces float64 anyway (see the docs). I hope that this covers everything. Otherwise, please let me know!

Good point. I should really check the stable documentation :)

So this PR LGTM.

@jotasi could you open some issue with the points mentioned there: https://github.com/scikit-learn/scikit-learn/pull/10527#issuecomment-361482040

jotasi commented:
I've opened an issue for the last point. I'll submit a PR for that if this PR gets merged. The other two should already be fixed by #10557 and #10569 respectively.

jotasi commented:
Sorry, that was not really true. Actually, the first point

> Adding numpy/scipy versions to circleci

is not really fixed, as the PR had to be reverted. It is not possible to simply add the minimum version of SciPy mentioned in the README (`0.13.3`) to circleCI, as one of the examples actually fails to work with this version (works only for `>0.14.0` as explained in this [comment](https://github.com/scikit-learn/scikit-learn/pull/10557#issuecomment-362510922) in the discussion to #10557). Should I open an issue for that as well as even though it does not seem to be easily fixable?

lesteve committed:
Minor tweaks

lesteve commented:
I pushed some minor tweaks, I will merge when CIs are green.

lesteve commented:
Circle is green and I checked that the examples looked fine. Merging thanks a lot @jotasi!
